db/virtual_table.cc

```diff
@@ -106,4 +106,91 @@ mutation_source memtable_filling_virtual_table::as_mutation_source() {
+            reader_permit permit,
+            const dht::partition_range& pr,
+            const query::partition_slice& slice,
+            const io_priority_class& pc,
+            tracing::trace_state_ptr trace_state,
+            streamed_mutation::forwarding fwd,
+            mutation_reader::forwarding fwd_mr) {
+
+        if (fwd == streamed_mutation::forwarding::yes) {
+            throw std::runtime_error("streamed_mutation::forwarding::yes unsupported in virtual tables");
+        }
+
+        if (fwd_mr == mutation_reader::forwarding::yes) {
+            throw std::runtime_error("mutation_reader::forwarding::yes unsupported in virtual tables");
+        }
+
+        // We cannot pass the partition_range directly to execute()
+        // because it is not guaranteed to be alive until execute() resolves.
+        // It is only guaranteed to be alive as long as the returned reader is alive.
+        // We achieve safety by mediating access through query_restrictions. When the reader
+        // dies, pr is cleared and execute() will get an exception.
+        struct my_result_collector : public result_collector, public query_restrictions {
+            queue_reader_handle handle;
+
+            // Valid until handle.is_terminated(), which is set to true when the
+            // queue_reader dies.
+            const dht::partition_range* pr;
+            mutation_reader::forwarding fwd_mr;
+
+            my_result_collector(schema_ptr s, reader_permit p, const dht::partition_range* pr, queue_reader_handle&& handle)
+                : result_collector(s, p)
+                , handle(std::move(handle))
+                , pr(pr)
+            { }
+
+            // result_collector
+            future<> take(mutation_fragment fragment) override {
+                return handle.push(std::move(fragment));
+            }
+
+            // query_restrictions
+            const dht::partition_range& partition_range() const override {
+                if (handle.is_terminated()) {
+                    throw std::runtime_error("read abandoned");
+                }
+                return *pr;
+            }
+        };
+
+        // FIXME: queue_reader doesn't support next_partition() needed for "distinct".
+        auto reader_and_handle = make_queue_reader(_s, permit);
+        auto consumer = std::make_unique<my_result_collector>(_s, permit, &pr, std::move(reader_and_handle.second));
+        auto f = execute(permit, *consumer, *consumer);
+        (void)f.then_wrapped([c = std::move(consumer)] (auto&& f) {
+            if (f.failed()) {
+                c->handle.abort(f.get_exception());
+            } else {
+                c->handle.push_end_of_stream();
+            }
+        });
+
+        auto rd = make_filtering_reader(std::move(reader_and_handle.first), pr, slice);
+
+        if (!_shard_aware) {
+            rd = make_filtering_reader(std::move(rd), [this] (const dht::decorated_key& dk) -> bool {
```

Comment:
Surely we don't want two filtering_readers that do different things.

Reply:
not part of this PR, explained below